File: sdk/AuthSys-CSHARP-Example/WinForms/src/AuthSys.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Runtime.InteropServices;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace AuthSys
{
    public class AuthSysException : Exception
    {
        public int StatusCode { get; }
        public string ErrorCode { get; }

        public AuthSysException(string message, int statusCode = 0, string errorCode = "") : base(message)
        {
            StatusCode = statusCode;
            ErrorCode = errorCode;
        }
    }

    public class AuthSysOptions
    {
        public string AppSecret { get; set; } = "";
        public string AppName { get; set; } = "";
        public string Version { get; set; } = "";
        public string ApiUrl { get; set; } = "https://api.authsys.dpdns.org/api/v1";
        public int TimeoutSeconds { get; set; } = 30;
        public int MaxRetries { get; set; } = 3;
        public bool SkipCertificateValidation { get; set; } = false;
        public bool EnableLogging { get; set; } = false;
    }

    public class AuthSys
    {
        private readonly AuthSysOptions _options;
        private readonly HttpClient _httpClient;
        private readonly HttpClientHandler _handler;
        private string _sessionToken = "";
        private bool _initialized = false;
        private Dictionary<string, object> _appVariables = new();
        private string _username = "";

        public AuthSys(AuthSysOptions options)
        {
            _options = options;
            _handler = new HttpClientHandler();
            if (options.SkipCertificateValidation)
            {
                _handler.ServerCertificateCustomValidationCallback = (message, cert, chain, errors) => true;
            }
            _httpClient = new HttpClient(_handler)
            {
                BaseAddress = new Uri(_options.ApiUrl),
                Timeout = TimeSpan.FromSeconds(_options.TimeoutSeconds)
            };
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }

        public AuthSys(string appSecret) : this(new AuthSysOptions { AppSecret = appSecret }) { }

        private void Log(string message)
        {
            if (_options.EnableLogging)
            {
                System.Diagnostics.Debug.WriteLine($"[AuthSys] {message}");
            }
        }

        private async Task<Dictionary<string, object>> SendRequestAsync(string endpoint, Dictionary<string, object> data = null, Dictionary<string, string> headers = null)
        {
            var url = $"client/{endpoint}";
            var json = data != null ? JsonSerializer.Serialize(data) : "{}";
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            for (int attempt = 0; attempt <= _options.MaxRetries; attempt++)
            {
                try
                {
                    Log($"POST {url} (attempt {attempt + 1})");
                    using var request = new HttpRequestMessage(HttpMethod.Post, url);
                    request.Content = content;
                    if (headers != null)
                    {
                        foreach (var h in headers)
                            request.Headers.Add(h.Key, h.Value);
                    }

                    var response = await _httpClient.SendAsync(request);
                    var responseBody = await response.Content.ReadAsStringAsync();
                    Log($"Response: {response.StatusCode} - {responseBody}");

                    if (!response.IsSuccessStatusCode)
                    {
                        var statusCode = (int)response.StatusCode;
                        string errorCode = "api_error";
                        switch (statusCode)
                        {
                            case 401: errorCode = "unauthorized"; break;
                            case 403: errorCode = "forbidden"; break;
                            case 404: errorCode = "not_found"; break;
                            case 429: errorCode = "rate_limited"; break;
                            case 503: errorCode = "maintenance"; break;
                        }
                        throw new AuthSysException(responseBody, statusCode, errorCode);
                    }

                    return JsonSerializer.Deserialize<Dictionary<string, object>>(responseBody) ?? new();
                }
                catch (AuthSysException)
                {
                    throw;
                }
                catch (Exception ex) when (attempt < _options.MaxRetries)
                {
                    Log($"Request error (attempt {attempt + 1}): {ex.Message}");
                    await Task.Delay(TimeSpan.FromMilliseconds(Math.Pow(2, attempt) * 1000));
                }
            }

            throw new AuthSysException("Request failed after all retries", 0, "network_error");
        }

        public async Task<Dictionary<string, object>> InitAsync()
        {
            Log("Initializing...");
            var data = new Dictionary<string, object>
            {
                ["app_secret"] = _options.AppSecret,
                ["version"] = _options.Version,
                ["app_name"] = _options.AppName,
                ["hwid"] = AuthSysHelpers.GetHwid()
            };

            var result = await SendRequestAsync("init", data);
            var status = result.GetValueOrDefault("status", "").ToString() ?? "";

            if (status == "update_required")
                throw new AuthSysException(result.GetValueOrDefault("message", "").ToString() ?? "Update required", 0, "version_mismatch");

            _initialized = status == "success" || status == "update_available";
            if (result.ContainsKey("variables"))
                _appVariables = JsonSerializer.Deserialize<Dictionary<string, object>>(JsonSerializer.Serialize(result["variables"])) ?? new();
            return result;
        }

        public async Task<Dictionary<string, object>> RegisterAsync(string username, string password, string licenseKey, string email = "")
        {
            if (!_initialized)
                throw new AuthSysException("Not initialized. Call InitAsync() first.", 0, "not_initialized");

            var data = new Dictionary<string, object>
            {
                ["app_secret"] = _options.AppSecret,
                ["username"] = username,
                ["password"] = password,
                ["license_key"] = licenseKey,
                ["hwid"] = AuthSysHelpers.GetHwid()
            };
            if (!string.IsNullOrEmpty(email))
                data["email"] = email;

            return await SendRequestAsync("register", data);
        }

        public async Task<Dictionary<string, object>> LoginAsync(string username, string password, int sessionLength = 86400)
        {
            if (!_initialized)
                throw new AuthSysException("Not initialized. Call InitAsync() first.", 0, "not_initialized");

            _sessionToken = "";
            var data = new Dictionary<string, object>
            {
                ["app_secret"] = _options.AppSecret,
                ["username"] = username,
                ["password"] = password,
                ["hwid"] = AuthSysHelpers.GetHwid(),
                ["session_length"] = sessionLength
            };

            var result = await SendRequestAsync("login", data);
            if (result.ContainsKey("token") && result["token"].ToString() != "")
                _sessionToken = result["token"].ToString() ?? "";
            if (result.ContainsKey("username"))
                _username = result["username"].ToString() ?? "";
            return result;
        }

        public async Task<Dictionary<string, object>> LicenseLoginAsync(string licenseKey, int sessionLength = 86400)
        {
            if (!_initialized)
                throw new AuthSysException("Not initialized. Call InitAsync() first.", 0, "not_initialized");

            _sessionToken = "";
            var data = new Dictionary<string, object>
            {
                ["app_secret"] = _options.AppSecret,
                ["license_key"] = licenseKey,
                ["hwid"] = AuthSysHelpers.GetHwid(),
                ["session_length"] = sessionLength
            };

            var result = await SendRequestAsync("license-login", data);
            if (result.ContainsKey("token") && result["token"].ToString() != "")
                _sessionToken = result["token"].ToString() ?? "";
            return result;
        }

        public async Task<Dictionary<string, object>> LicenseCheckAsync(string licenseKey)
        {
            var data = new Dictionary<string, object>
            {
                ["app_secret"] = _options.AppSecret,
                ["license_key"] = licenseKey
            };
            return await SendRequestAsync("license/check", data);
        }

        public async Task<Dictionary<string, object>> VerifyAsync()
        {
            if (string.IsNullOrEmpty(_sessionToken))
                throw new AuthSysException("No active session. Login first.", 0, "no_session");

            var headers = new Dictionary<string, string>
            {
                ["Authorization"] = $"Bearer {_sessionToken}",
                ["X-HWID"] = AuthSysHelpers.GetHwid()
            };
            return await SendRequestAsync("verify", null, headers);
        }

        public async Task<Dictionary<string, object>> SendChatMessageAsync(int roomId, string message)
        {
            if (string.IsNullOrEmpty(_sessionToken))
                throw new AuthSysException("No active session. Login first.", 0, "no_session");

            var headers = new Dictionary<string, string>
            {
                ["Authorization"] = $"Bearer {_sessionToken}"
            };
            var endpoint = $"chat/send?room_id={roomId}&message={Uri.EscapeDataString(message)}";
            return await SendRequestAsync(endpoint, null, headers);
        }

        public async Task<Dictionary<string, object>> RegisterDeviceAsync(string hwid, string deviceName = "")
        {
            var data = new Dictionary<string, object>
            {
                ["app_secret"] = _options.AppSecret,
                ["hwid"] = hwid
            };
            if (!string.IsNullOrEmpty(deviceName))
                data["device_name"] = deviceName;
            return await SendRequestAsync("device/register", data);
        }

        public async Task<Dictionary<string, object>> CheckDeviceAsync(string hwid)
        {
            var data = new Dictionary<string, object>
            {
                ["app_secret"] = _options.AppSecret,
                ["hwid"] = hwid
            };
            return await SendRequestAsync("device/check", data);
        }

        public object GetVariable(string key)
        {
            return _appVariables.ContainsKey(key) ? _appVariables[key] : null;
        }

        public Dictionary<string, object> GetAllVariables()
        {
            return _appVariables;
        }

        public void Logout()
        {
            _sessionToken = "";
        }

        public bool IsAuthenticated => !string.IsNullOrEmpty(_sessionToken);
        public bool IsInitialized => _initialized;
        public string Username => _username;
    }
}

File: sdk/AuthSys-CPP-Example/AuthSys.cpp
#include "AuthSys.hpp"
#include <curl/curl.h>
#include <sstream>
#include <thread>
#include <chrono>
#include <cstdlib>
#include <cstdio>
#include <cstring>

#ifdef _WIN32
#include <windows.h>
#include <direct.h>
#else
#include <unistd.h>
#endif

static size_t WriteCallback(void* contents, size_t size, size_t nmemb, void* userp) {
    ((std::string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}

std::string getHwid() {
#ifdef _WIN32
    FILE* pipe = _popen("wmic csproduct get uuid", "r");
    if (pipe) {
        char buffer[256];
        std::string result;
        while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
            std::string line(buffer);
            size_t pos = line.find_first_not_of(" \t\r\n");
            if (pos != std::string::npos && line.substr(pos) != "UUID" && !line.substr(pos).empty()) {
                result = line.substr(pos);
                size_t end = result.find_last_not_of(" \t\r\n");
                if (end != std::string::npos) result = result.substr(0, end + 1);
                _pclose(pipe);
                return result;
            }
        }
        _pclose(pipe);
    }
    return "UNKNOWN_HWID";
#elif __linux__
    FILE* f = fopen("/etc/machine-id", "r");
    if (f) {
        char buf[256];
        if (fgets(buf, sizeof(buf), f)) {
            fclose(f);
            std::string result(buf);
            size_t end = result.find_last_not_of(" \t\r\n");
            if (end != std::string::npos) result = result.substr(0, end + 1);
            return result;
        }
        fclose(f);
    }
    return "UNKNOWN_HWID";
#elif __APPLE__
    FILE* pipe = popen("ioreg -rd1 -c IOPlatformExpertDevice | grep IOPlatformUUID", "r");
    if (pipe) {
        char buffer[512];
        std::string result;
        while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
            std::string line(buffer);
            size_t start = line.find('"');
            if (start != std::string::npos) {
                size_t end = line.find('"', start + 1);
                if (end != std::string::npos) {
                    result = line.substr(start + 1, end - start - 1);
                    pclose(pipe);
                    return result;
                }
            }
        }
        pclose(pipe);
    }
    return "UNKNOWN_HWID";
#else
    return "UNKNOWN_HWID";
#endif
}

void AuthSys::log(const std::string& message) {
    if (_options.enableLogging) {
        std::cout << "[AuthSys] " << message << std::endl;
    }
}

std::string AuthSys::sendRequest(const std::string& endpoint, const std::string& jsonPayload,
                                  const std::map<std::string, std::string>& headers) {
    std::string url = _options.apiUrl + "/client/" + endpoint;
    std::string lastError;

    for (int attempt = 0; attempt <= _options.maxRetries; attempt++) {
        log("POST " + url + " (attempt " + std::to_string(attempt + 1) + ")");

        CURL* curl = curl_easy_init();
        if (!curl) {
            lastError = "Failed to initialize curl";
            continue;
        }

        std::string responseBody;
        struct curl_slist* headerList = nullptr;

        headerList = curl_slist_append(headerList, "Content-Type: application/json");
        headerList = curl_slist_append(headerList, "Accept: application/json");
        for (const auto& h : headers) {
            std::string header = h.first + ": " + h.second;
            headerList = curl_slist_append(headerList, header.c_str());
        }

        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headerList);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &responseBody);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, _options.timeout);
        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, _options.timeout);

        if (!jsonPayload.empty()) {
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonPayload.c_str());
        }

        if (_options.skipCertificateValidation) {
            curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
            curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
        }

        CURLcode res = curl_easy_perform(curl);
        long httpCode = 0;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &httpCode);

        curl_slist_free_all(headerList);
        curl_easy_cleanup(curl);

        if (res != CURLE_OK) {
            lastError = curl_easy_strerror(res);
            log("Request error (attempt " + std::to_string(attempt + 1) + "): " + lastError);
            if (attempt < _options.maxRetries) {
                std::this_thread::sleep_for(std::chrono::milliseconds(1 << attempt * 1000));
            }
            continue;
        }

        log("Response: " + std::to_string(httpCode) + " - " + responseBody);

        if (httpCode < 200 || httpCode >= 300) {
            handleHttpError((int)httpCode, responseBody);
        }

        return responseBody;
    }

    throw AuthSysException(lastError.empty() ? "Request failed after all retries" : lastError, 0, "network_error");
}

void AuthSys::handleHttpError(int statusCode, const std::string& responseBody) {
    std::string errorCode = "api_error";
    switch (statusCode) {
        case 401: errorCode = "unauthorized"; break;
        case 403: errorCode = "forbidden"; break;
        case 404: errorCode = "not_found"; break;
        case 429: errorCode = "rate_limited"; break;
        case 503: errorCode = "maintenance"; break;
    }
    throw AuthSysException(responseBody, statusCode, errorCode);
}

AuthSys::AuthSys(const AuthSysOptions& options)
    : _options(options), _sessionToken(""), _initialized(false), _username("") {
    curl_global_init(CURL_GLOBAL_DEFAULT);
}

void AuthSys::init() {
    log("Initializing...");
    std::string json = "{\"app_secret\":\"" + _options.appSecret +
                       "\",\"version\":\"" + _options.version +
                       "\",\"app_name\":\"" + _options.appName +
                       "\",\"hwid\":\"" + getHwid() + "\"}";

    std::string res = sendRequest("init", json, {});

    if (res.find("\"update_required\"") != std::string::npos) {
        throw AuthSysException("Update required", 0, "version_mismatch");
    }

    _initialized = res.find("\"success\"") != std::string::npos ||
                   res.find("\"update_available\"") != std::string::npos;
}

std::string AuthSys::registerUser(const std::string& username, const std::string& password,
                                   const std::string& licenseKey, const std::string& email) {
    if (!_initialized) {
        throw AuthSysException("Not initialized. Call init() first.", 0, "not_initialized");
    }

    std::string json = "{\"app_secret\":\"" + _options.appSecret +
                       "\",\"username\":\"" + username +
                       "\",\"password\":\"" + password +
                       "\",\"license_key\":\"" + licenseKey +
                       "\",\"hwid\":\"" + getHwid() + "\"}";

    if (!email.empty()) {
        json = json.substr(0, json.size() - 1) + ",\"email\":\"" + email + "\"}";
    }

    return sendRequest("register", json, {});
}

std::string AuthSys::login(const std::string& username, const std::string& password, int sessionLength) {
    if (!_initialized) {
        throw AuthSysException("Not initialized. Call init() first.", 0, "not_initialized");
    }

    _sessionToken = "";
    std::string json = "{\"app_secret\":\"" + _options.appSecret +
                       "\",\"username\":\"" + username +
                       "\",\"password\":\"" + password +
                       "\",\"hwid\":\"" + getHwid() +
                       "\",\"session_length\":" + std::to_string(sessionLength) + "}";

    std::string res = sendRequest("login", json, {});

    size_t tokenPos = res.find("\"token\":\"");
    if (tokenPos != std::string::npos) {
        tokenPos += 9;
        size_t endPos = res.find("\"", tokenPos);
        if (endPos != std::string::npos) {
            _sessionToken = res.substr(tokenPos, endPos - tokenPos);
        }
    }

    return res;
}

std::string AuthSys::licenseLogin(const std::string& licenseKey, int sessionLength) {
    if (!_initialized) {
        throw AuthSysException("Not initialized. Call init() first.", 0, "not_initialized");
    }

    _sessionToken = "";
    std::string json = "{\"app_secret\":\"" + _options.appSecret +
                       "\",\"license_key\":\"" + licenseKey +
                       "\",\"hwid\":\"" + getHwid() +
                       "\",\"session_length\":" + std::to_string(sessionLength) + "}";

    std::string res = sendRequest("license-login", json, {});

    size_t tokenPos = res.find("\"token\":\"");
    if (tokenPos != std::string::npos) {
        tokenPos += 9;
        size_t endPos = res.find("\"", tokenPos);
        if (endPos != std::string::npos) {
            _sessionToken = res.substr(tokenPos, endPos - tokenPos);
        }
    }

    return res;
}

std::string AuthSys::licenseCheck(const std::string& licenseKey) {
    std::string json = "{\"app_secret\":\"" + _options.appSecret +
                       "\",\"license_key\":\"" + licenseKey + "\"}";
    return sendRequest("license/check", json, {});
}

std::string AuthSys::verify() {
    if (_sessionToken.empty()) {
        throw AuthSysException("No active session. Login first.", 0, "no_session");
    }

    std::map<std::string, std::string> headers;
    headers["Authorization"] = "Bearer " + _sessionToken;
    headers["X-HWID"] = getHwid();

    return sendRequest("verify", "", headers);
}

std::string AuthSys::sendChatMessage(int roomId, const std::string& message) {
    if (_sessionToken.empty()) {
        throw AuthSysException("No active session. Login first.", 0, "no_session");
    }

    std::map<std::string, std::string> headers;
    headers["Authorization"] = "Bearer " + _sessionToken;

    std::string endpoint = "chat/send?room_id=" + std::to_string(roomId) +
                           "&message=" + message;

    return sendRequest(endpoint, "", headers);
}

std::string AuthSys::registerDevice(const std::string& hwid, const std::string& deviceName) {
    std::string json = "{\"app_secret\":\"" + _options.appSecret +
                       "\",\"hwid\":\"" + hwid + "\"}";

    if (!deviceName.empty()) {
        json = json.substr(0, json.size() - 1) + ",\"device_name\":\"" + deviceName + "\"}";
    }

    return sendRequest("device/register", json, {});
}

std::string AuthSys::checkDevice(const std::string& hwid) {
    std::string json = "{\"app_secret\":\"" + _options.appSecret +
                       "\",\"hwid\":\"" + hwid + "\"}";
    return sendRequest("device/check", json, {});
}

std::string AuthSys::getVariable(const std::string& key) const {
    auto it = _appVariables.find(key);
    return it != _appVariables.end() ? it->second : "";
}

std::map<std::string, std::string> AuthSys::getAllVariables() const {
    return _appVariables;
}

void AuthSys::logout() {
    _sessionToken = "";
}

bool AuthSys::isAuthenticated() const {
    return !_sessionToken.empty();
}

bool AuthSys::isInitialized() const {
    return _initialized;
}

std::string AuthSys::getUsername() const {
    return _username;
}
